### PR TITLE
Add parity versions of integration tests to the CI builds.

### LINKED ISSUE #453: Support Parity in the CI
## Problem Definition

Parity is not used in the continuous integration.

## Solution

Add `parity` to Raiden's pytest fixtures as a `blockchain type` and use it in travis.

## Tasklist

- [ ] Create the parity blockchain type and all the required fixtures for configuring, starting, and cleanup the private blockchain.
- [ ] Add to the travis matrix

### PR DESCRIPTION
**Pull request by jomuel**
Fixes #453

### CODE REVIEW

**Comment by jomuel** [comment]
The skipped tests are tracked by issue #3565

File: raiden/tests/conftest.py [hunk 1]
```diff
@@ -23,7 +23,7 @@
 def pytest_addoption(parser):
     parser.addoption(
         '--blockchain-type',
-        choices=['geth'],
+        choices=['geth', 'parity'],
```
Comment (LefterisJP) [inline raiden/tests/conftest.py]: This list/tuple is used in other places too. Like [here](https://github.com/raiden-network/raiden/pull/3483/files#diff-8e4c09e9a7222a70cd9463af4c512a23R51). Shouldn't we put it in the settings.py as `SUPPORTED_CLIENTS` so that it's hard-coded in a common place?

File: raiden/tests/integration/fixtures/blockchain.py [hunk 2]
```diff
@@ -46,65 +48,67 @@ def web3(
...
+    rpc_port = blockchain_rpc_ports[0]
+    endpoint = f'http://{host}:{rpc_port}'
+    web3 = Web3(HTTPProvider(endpoint))
+
+    assert len(blockchain_private_keys) == len(blockchain_rpc_ports)
+    assert len(blockchain_private_keys) == len(blockchain_p2p_ports)
+
+    eth_nodes = [
+        EthNodeDescription(
+            key,
```
Comment (LefterisJP) [inline raiden/tests/integration/fixtures/blockchain.py]: please use kwargs here. I know you just moved the code but it's a good chance to add them :)

File: raiden/tests/integration/fixtures/raiden_network.py [hunk 3]
```diff
@@ -16,6 +16,15 @@
 from raiden.tests.utils.tests import shutdown_apps_and_cleanup_tasks
 
 
+def wait_for_min_blocks(raiden, web3):
```
Comment (LefterisJP) [inline raiden/tests/integration/fixtures/raiden_network.py]: I don't think this function is needed. We already got something similar in the `waiting.py` module [here](https://github.com/raiden-network/raiden/blob/a6e95b986892fb8185a14ca91a093eb847b71b0d/raiden/waiting.py#L19)

File: raiden/tests/integration/fixtures/raiden_network.py [hunk 4]
```diff
@@ -85,7 +96,8 @@ def raiden_chain(
         )
 
     exception = RuntimeError('`raiden_chain` fixture setup failed, token networks unavailable')
-    with gevent.Timeout(seconds=30, exception=exception):
+    timeout = 120 if blockchain_type == 'parity' else 30
```
Comment (LefterisJP) [inline raiden/tests/integration/fixtures/raiden_network.py]: Since 

`timeout = 120 if blockchain_type == 'parity' else 30`

is used in multiple places in these tests I would suggest putting it in a function along with a docstring explaining why the choice of different timeouts per client

File: raiden/tests/integration/long_running/test_integration_events.py [hunk 5]
```diff
@@ -339,15 +340,17 @@ def test_query_events(
...
-        chain=app0.raiden.chain,
-        token_network_registry_address=app0.raiden.default_registry.address,
-        contract_manager=contract_manager,
-        events=ALL_EVENTS,
-        from_block=999999998,
-        to_block=999999999,
-    )
-    assert not events
+    if blockchain_type == 'geth':
+        # TODO why do we query events in nonexisting blocks here?
```
Comment (LefterisJP) [inline raiden/tests/integration/long_running/test_integration_events.py]: The test wants to confirm that querying for non-existing blocks returns an empty list.

File: tools/startcluster.py [hunk 6]
```diff
@@ -40,7 +40,7 @@ def main():
         p2p_port = START_PORT + i
         rpc_port = START_RPCPORT + i
 
-        description = GethNodeDescription(
+        description = EthNodeDescription(
             node_key,
```
Comment (LefterisJP) [inline tools/startcluster.py]: can you add kwargs here?

File: raiden/tests/utils/eth_node.py [hunk 7]
```diff
@@ -189,22 +242,68 @@ def geth_init_datadir(datadir: str, genesis_path: str):
...
+    iv = keyfile_json['crypto']['cipherparams']['iv']
+    keyfile_json['crypto']['cipherparams']['iv'] = f'{iv:0>32}'
+    # Parity expects a string of length 32 here, but eth_keyfile does not pad
+    with open(f'{path}/keyfile', 'w') as keyfile:
+        json.dump(keyfile_json, keyfile)
+
+    return path
+
+
+def parity_create_account(node_configuration, base_path, chain_spec):
```
Comment (LefterisJP) [inline raiden/tests/utils/eth_node.py]: can you add types in the args?

File: raiden/tests/utils/eth_node.py [hunk 8]
```diff
@@ -189,22 +242,68 @@ def geth_init_datadir(datadir: str, genesis_path: str):
...
+
+    return_code = process.wait()
+    if return_code:
+        raise RuntimeError(
+            f'Creation of parity signer account failed with return code {return_code}',
+        )
+
+
+def eth_wait_and_check(
         web3,
```
Comment (LefterisJP) [inline raiden/tests/utils/eth_node.py]: add types here

File: raiden/tests/utils/eth_node.py [hunk 9]
```diff
@@ -419,24 +526,38 @@ def geth_run_private_blockchain(
...
-        accounts_to_fund,
-        seal_account,
-        random_marker,
-    )
+    if blockchain_type == 'geth':
+        eth_node_config_set_bootnodes(nodes_configuration)
+
+        genesis_path = os.path.join(base_datadir, 'custom_genesis.json')
+        geth_generate_poa_genesis(
+            genesis_path,
```
Comment (LefterisJP) [inline raiden/tests/utils/eth_node.py]: here and below please add kwargs

File: raiden/tests/integration/fixtures/raiden_network.py (hunk 3 above)
Comment (jomuel) [reply raiden/tests/integration/fixtures/raiden_network.py]: `waiting.wait_for_blocks` cannot be used here. We need to wait for the chain to grow long enough before starting the apps, so we don't have a `RaidenService` instance to pass to it yet.

Comment (LefterisJP) [reply raiden/tests/integration/fixtures/raiden_network.py]: Hmm if we don't have a Raiden instance then what is the `raiden` argument for? Which reminds me ... put the types of the arguments ^_^


Also in any case this should probably be in the waiting.py ?

Comment (jomuel) [reply raiden/tests/integration/fixtures/raiden_network.py]: Good point, I was confused here. We have a `RaidenService` object at this point, but its `wal` attribute is not yet initialized. That's why `wait_for_block` will not work.

File: raiden/tests/integration/fixtures/blockchain.py [hunk 10]
```diff
@@ -17,6 +21,8 @@
 # pylint: disable=redefined-outer-name,too-many-arguments,unused-argument,too-many-locals
 
 _GETH_LOGDIR = os.environ.get('RAIDEN_TESTS_GETH_LOGSDIR', False)
+_GETH_DATADIR = os.environ.get('RAIDEN_TESTS_GETH_DATADIR', False)
```
Comment (konradkonrad) [inline raiden/tests/integration/fixtures/blockchain.py]: hmm, I don't see why the default return value `False` would be necessary -- `None` should be fine for all later `if OS_VAR:...` conditionals. Can you remove it for both `_GETH_...` variables?

Comment (jomuel) [reply raiden/tests/integration/fixtures/blockchain.py]: I just realized that the DATADIR variables are not used anymore anyway, so I removed them (and the `False` default for `_GETH_LOGDIR`).

Comment (konradkonrad) [reply raiden/tests/integration/fixtures/blockchain.py]: :+1:

File: raiden/tests/integration/fixtures/raiden_network.py [hunk 11]
```diff
@@ -14,6 +14,21 @@
     wait_for_token_networks,
 )
 from raiden.tests.utils.tests import shutdown_apps_and_cleanup_tasks
+from raiden.waiting import wait_for_block_using_web3
+
+
+def wait_for_confirmed_block(blockchain_services, raiden_apps):
+    wait_for_block_using_web3(
+        web3=blockchain_services[2].client.web3,
```
Comment (konradkonrad) [inline raiden/tests/integration/fixtures/raiden_network.py]: why are you using `[2]` here / how are you sure that index is actually defined? maybe you want to use `[-1]` ?

File: raiden/tests/integration/long_running/test_integration_events.py [hunk 12]
```diff
@@ -339,15 +340,18 @@ def test_query_events(
...
-        chain=app0.raiden.chain,
-        token_network_registry_address=app0.raiden.default_registry.address,
-        contract_manager=contract_manager,
-        events=ALL_EVENTS,
-        from_block=999999998,
-        to_block=999999999,
-    )
-    assert not events
+    if blockchain_type == 'geth':
+        # FIXME: This is apparently meant to verifiy that querying nonexisting blocks
```
Comment (konradkonrad) [inline raiden/tests/integration/long_running/test_integration_events.py]: `s/verifiy/verify` :D

File: raiden/tests/integration/network/transport/test_udp.py [hunk 13]
```diff
@@ -69,7 +69,7 @@ def test_udp_unreachable_node(raiden_network, skip_if_not_udp):
 @pytest.mark.parametrize('number_of_nodes', [1])
 @pytest.mark.parametrize('channels_per_node', [0])
 @pytest.mark.parametrize('number_of_tokens', [1])
-def test_suite_survives_unhandled_exception(raiden_network):
+def test_suite_survives_unhandled_exception(raiden_network, skip_if_parity):
```
Comment (konradkonrad) [inline raiden/tests/integration/network/transport/test_udp.py]: `skip_if_parity` should be a marker imho

File: raiden/tests/integration/transfer/test_refundtransfer.py [hunk 14]
```diff
@@ -457,15 +458,16 @@ def patched_update(raiden, event):
...
+    with gevent.Timeout(timeout):
         unlock_app0 = wait_for_state_change(
             app0.raiden,
             ContractReceiveChannelBatchUnlock,
             {'participant': app0.raiden.address},
             retry_timeout,
         )
     assert unlock_app0.returned_tokens == 50
-    with gevent.Timeout(10):
+    with gevent.Timeout(timeout):
```
Comment (konradkonrad) [inline raiden/tests/integration/transfer/test_refundtransfer.py]: couldn't you re-use `timeout(blockchain_type)` as above?

File: raiden/tests/integration/network/transport/test_udp.py (hunk 13 above)
Comment (jomuel) [reply raiden/tests/integration/network/transport/test_udp.py]: Same problem as with `skip_if_not_udp` and `skip_if_not_matrix` - a marker has no access to the fixtures and cannot see the blockchain type.

File: raiden/tests/integration/transfer/test_refundtransfer.py (hunk 14 above)
Comment (jomuel) [reply raiden/tests/integration/transfer/test_refundtransfer.py]: I think it's more readable this way, than calling a function in a different module. I thought about parametrizing `timeout(blockchain_type)` for different timeouts, but the 10 second case doesn't happen that often.

File: raiden/tests/integration/fixtures/raiden_network.py (hunk 11 above)
Comment (jomuel) [reply raiden/tests/integration/fixtures/raiden_network.py]: `blockchain_services` is a `NamedTuple`, it always has 4 items. It would be better to use the attribute name of course, so I changed it to `blockchain_services.deploy_service.client.web3`.

File: raiden/tests/integration/fixtures/raiden_network.py [hunk 15]
```diff
@@ -14,6 +14,21 @@
     wait_for_token_networks,
 )
 from raiden.tests.utils.tests import shutdown_apps_and_cleanup_tasks
+from raiden.waiting import wait_for_block_using_web3
+
+
+def wait_for_confirmed_block(blockchain_services, raiden_apps):
```
Comment (hackaugusto) [inline raiden/tests/integration/fixtures/raiden_network.py]: Why is this necessary?
